Restructure IPCache to handle metadata merging

joestringer opened the pull request:
Meta: https://github.com/cilium/cilium/issues/21142

IPCache has historically been based around a concept of associating IP prefixes with a certain _Identity_, where each caller independently determines the exact identity that a IP Prefix should have. Each of these callers has a corresponding `source.Source` which decides the priority between these different sources of Identity.

More recently however, features like #14724 have driven the need to combine multiple sources of information, such as associating both the `remote-node` labels (from k8s `CiliumNode` resources) and `kube-apiserver` labels (from k8s `Endpoints` / `EndpointSlices` resources). Having an exclusionary policy where only one source of information can be correct can have implications like causing conflicts between policies that match on information from each source. Up until now, we've resolved these conflicts somewhat manually by forcing the callers to handle the merging of this information. For most callers, they're independent enough that users would only encounter bugs in rare conditions of using two features together in an unexpected way. We have made some attempts in certain paths to resolve these conflicts where necessary, including the example above. However, over time this code becomes more and more :spaghetti: with each caller attempting to resolve conflicts with the other callers.

This PR reworks the metadata map inside the IPCache to handle multiple sources of information at once, keying them in the map first by IP prefix, then by a `{resourceType, resourceNamespace, resourceName}` tuple (`ResourceID`). When merging the information, a new `prefixInfo` structure provides convenient merging of the labels from all of the sources of information. In effect, this introduced one additional layer of indirection with a series of chained structures represented in the mermaid chart below:

```mermaid
flowchart
  subgraph resourceInfo
  labels.Labels
  source.Source
  end
  subgraph prefixInfo
  UA[ResourceID]-->LA[resourceInfo]
  UB[ResourceID]-->LB[resourceInfo]
  ...
  end
  subgraph identityMetadata
  IP_Prefix-->prefixInfo
  end
```

Furthemore, rather than the previous implementation where _adds_ into the metadata map were asynchronous and _deletes_ were synchronous, this PR refactors the code to force both paths to be asynchronous. The first step which is synchronous is to update the desired state of the IPCache metadata, ie upserting the labels into the metadata map. This is called directly from various watchers like k8s resource watchers or the node manager. These callers then `TriggerLabelInjection()` to apply the changes out-of-band. This causes a goroutine to separately iterate through the changes that need to be made, then ensure that the control plane policy engine (in the `SelectorCache`) to be updated, followed by the dataplane policy maps, then finally followed by the dataplane ipcache. This ensures that policy transitions for newly allocated identities will occur in the correct order to prevent packet drops when different labels are associated with (or removed from) IP prefixes.

By making the deletes also asynchronous and combining the core logic, we are able to also delete a bunch of code, and more consistently handle the various cases of changing the sets of labels associated with prefixes.

See the individual commits for more details.

Combined design and implementation with @christarazi .

Supersedes: https://github.com/cilium/cilium/pull/19758
Related: https://github.com/cilium/cilium/issues/18301

joestringer committed:
ipcache/md: Rework metadata cache for any labels

Rework the internals of the "IDMD" Identity <-> Metadata cache so that
it can store multiple pieces of information coming from different
sources, rather than only labels coming from the kube-apiserver identity
feature. Initially in this commit, this only includes one feature source
which only stores labels internally, which is what's necessary for the
current user of this code.

In future, we anticipate extending the core info structure here to
include not just labels, but also auxiliary data like port name mappings
and encryption indexes shared by other nodes.

Information is keyed by the ResourceID, typically a {resource, name,
namespace} tuple of a resource from kubernetes, which allows convenient
insertion/deletion, and provides a way to trace the source of the
information for debugging purposes. This can mean that the same
information may be stored multiple times by different ResourceID inside
the structure, but this is not expected to be the normal case so the
memory cost for this is expected to be minimal.

Co-authored-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache/md: Add queue to handle identity updates

Add a queue into the metadata structure for handling the set of prefixes
which need to be processed as part of the next label injection trigger.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache/md: Update label injection diagram

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache/md: Support incremental ipcache updates

Previously, the InjectLabels() function which reacts to updates to the
IPCache metadata would iterate through the entire metadata map in order
to detect changes that occur, and subsequently react to those updates.
The reaction includes potentially allocating new identities, then
triggering updates into the datapath policymaps and the datapath's ipcache.

Given that there was no way to track which prefixes had been updated,
this meant iterating over the entire map, perhaps re-allocating
identities for prefixes that had not changed. This also made it more
difficult in future to support different types of updates to the ipcache
from other sources, since each feature would need to have special logic
encoded in this function to correctly handle the identity allocation.

Rework the InjectLabels() implementation to handle a queue of prefixes
to update instead. This commit only modifies the path used while
associating new sets of labels with a given prefix. A follow up commit
will handle the delete cases. However, given that adding new sets of
labels could also cause an existing prefix -> Labels (and hence Identity)
mapping to change, even this case must handle the cleanup of old
identities, which is operationally similar to deleting labels (and hence
identities).

This new code is also built around asynchronous updates to the label
associations for the prefixes. When triggered it only tracks which
prefixes have been changed, and not yet what it must do to ensure that
these prefixes have the correct identity. It uses two key pieces of
information to determine what must be done in order to ensure the
correct association:
* The desired state of the labels, prepared in the IPCache.metadata map
  during a previous UpsertMetadata() call, and
* The realized state of the ipcache itself.

The first time that a set of labels is associated with a prefix, the
ipcache will likely not have any corresponding entry, so this case is
simple: Check the desired state of the labels, allocate a corresponding
identity, and then update the policy selectorcache, datapath policymaps,
and finally the ipcache.

The second time that a (new) set of labels is associated with a prefix,
it gets a bit more complicated: The previous set of labels may still be
relevant. First, a lookup into the existing ipcache tells the function
that this prefix is already associated with some set of labels.

* First, determine the complete set of labels that should be associated
  with this prefix. In this commit, the new labels are likely a
  superset of the existing set. A subsequent commit will handle disjoint
  sets.
* Allocate a new identity corresponding to this complete set of labels.
* Update the SelectorCache and endpoint policy maps to potentially allow
  this new identity. Traffic will not yet be identified with this new
  identity, but this prepares the datapath policy engine for that
  eventuality.
* Update the IPcache to identify this prefix as having the new identity.
* Now that the new identity is in use in the datapath, we can remove
  references to the old identity (in the SelectorCache and policymaps)
* Finally, given that the IPCache no longer associates this prefix with
  the old identity, release the reference on the old identity to ensure
  that the old identity is returned back to the identity allocation pool.

Previously, the "identity release" logic was performed as part of the
core loop, and primarily based on whether any allocated identity was
considered "new" or had the "kube-apiserver" labels associated with it,
which was a bit difficult to reason about. That code is now simpler,
being tracked via an `idsToDelete` map during the processing of modified
prefixes, and the release is moved to the end of the function to ensure
that it is not prematurely released.

Co-authored-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache/md: Derive context for updates from controller

Most actual ipcache updates will in future be occurring as a result of
triggering the label injection controller, so propagate the context from
that controller down into the various child functions where they need a
context for identity allocation and datapath map updates.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache/md: Derive metadata source from Labels

Now that each individual PrefixInfo (labels + source) is held together
in one big happy identityMetadata map, we can rely directly on the map
for the correct source for updating the IPCache, rather than forcing
each caller of the trigger to pass it down.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache/md: Handle deletions asynchronously

Now that the IPCache handles metadata updates incrementally, we can
switch the deletion path over to use this same incremental update logic.
This allows the users of the deletion APIs to inject updates into the
ipcache in a similar manner to the users of the add APIs, and have the
updates incrementally triggered into the subsequent subsystems (policy,
datapath) consistently. This means that updates into the ipcache from
both add and delete paths for the kube-apiserver policy feature will
only actually occur from a single goroutine, triggered by the
TriggerLabelInjection call.

This removes the need to reason about multiple concurrent adds or
deletions into the IPcache occurring from the kube-apiserver policy
feature, and it also lays the path to do the same in future for CIDR
policy and other subsystems. This way, each user of the IPCache can
propagate the information that it intends to push into the IPCache, then
the IPCache itself can decide how to handle those updates and how to
combine information from the various subsystems.

Some key observations here:
* The previous commit actually includes 90% of the logic required to
  implement deletions, based partially on the previous 'add' code and
  partially on the code being deleted here.
* What changes here is adding support in InjectLabels() for the case
  where a set of labels is removed from a prefix. If all labels are
  removed, then this results in the 'IPCache.metadata' map having no
  corresponding entry, so in this case the corresponding old identity
  currently in 'IPCache' should be removed from the selectorcache,
  policymaps, and the IPCache.
  * Caveat: This is only the case if _only_ the metadata map has
    references to the identity. At this point in time, CIDR policies for
    instance are not yet converted over to the metadata map approach for
    associating labels with prefixes, so that path may independently
    allocate their own identities. If those are still referenced from
    CIDR policy, then the label injector should simply remove references
    to the corresponding identities but not remove it from the ipcache.
* Another case is when there are some set of labels (eg A, B) associated
  with a prefix, then one set (eg B) is removed. The result is that a
  previous identity with labels (A, B) must be removed, and a new
  identity with labels (A) should be allocated / associated with the
  prefix. In general, this is very similar to the existing case where a
  set of labels is expanded by associating new labels with the prefix
  (already handled in the previous commit).
  * This also has a curly case: Each set of labels has a source
    associated, for instance initially there could be "remote-node"
    (source: custom-resource) and "kube-apiserver" (source:
    kube-apiserver). When previously upserting into the IPCache, the
    source will be kube-apiserver. If the kube apiserver is no longer
    associated with the IP, and hence that label removed, then the
    resulting set of labels will only be "remote-node", with source
    "custom-resource". Given that the source "custom-resource" has a
    lower priority in pkg/source than kube-apiserver source, we cannot
    update the ipcache directly with the new set of labels using the
    "custom-resource" source. However, the label removal is still
    legitimate. To work around the clunky APIs, the function here just
    overrides the source check in the IPCache.Upsert(). We should be
    able to remove this clunkiness over time when the metadata map is
    the primary source of information for prefixes, but more refactoring
    is necessary to get to that point.
* Now that the label removal doesn't have its own independent logic from
  a separate goroutine, there is no longer a need to use the
  'applyChangesMU' mutex in the metadata cache to ensure safety around
  the critical section. Furthermore, the core InjectLabels() call
  doesn't modify the metadata map. So, we can remove one lock and reduce
  the other lock down to a read mutex rather than holding it for write.

Co-authored-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache/md: Move host label logic to injectLabels

Move the logic to change the host label into injectLabels() so that the
caller doesn't need to worry about this detail. This allows injectLabels
to fully handle ensuring that the specified labels are present in the
returned identity, which will simplify further upcoming changes to
generalize this code. In particular it will make it easier to handle
identity (re-)allocations for other label-based features.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache/md: Make IPCache.metadata.remove() generic

Back in the commit "ipcache/md: Rework metadata cache for any labels",
the underlying mechanism for storing IP metadata was made more generic
to store any information that 'IPMetadata' represents. At the time, the
upsert codepaths were made generic to handle different types of incoming
information, but the remove paths remained specific to labels.

Now that the remove paths have been tidied up, we can more easily
abstract out the remove path as well. This should mean that future
patches which wish to store alternative types of information in the
ipcache should only need to update pkg/ipcache/types.go for the new data
type, then pull that information out in the main InjectLabels() loop in
order to create the new IPCache entry using the new info.

Signed-off-by: Joe Stringer <joe@cilium.io>

christarazi commented on pkg/ipcache/metadata.go:
```diff
@@ -69,11 +69,41 @@ type metadata struct {
+	return
+}
+
+func (m *metadata) enqueuePrefixUpdates(prefixes []string) {
```
Minor nit: We can slightly improve this API by supporting a variadic param instead of a raw slice. The body of this function wouldn't actually need to change. This allows calls to this function to simply be `enqueuePrefixUpdates(prefix1, prefix2, ...)` without having to write the _slightly_ clunky `enqueuePrefixUpdates([]string{prefix1, ...})`.

christarazi commented on pkg/ipcache/metadata.go:
```diff
@@ -157,111 +165,155 @@ func (m *metadata) get(prefix string) prefixInfo {
+		if entryExists {
+			// 'prefix' is being removed or modified, so some
+			// previous iteration of this code hit one the
+			// 'injectLabels' case above, thereby allocating a
```
Nit: reads slightly off, "code hit one the"

christarazi commented on pkg/ipcache/metadata.go:
```diff
@@ -157,111 +165,155 @@ func (m *metadata) get(prefix string) prefixInfo {
+			log.WithError(err).WithFields(logrus.Fields{
+				logfields.IPAddr:   ip,
+				logfields.Identity: id,
+			}).Error("Failed to replace ipcache entry with new identity after label removal. Traffic may be disrupted.")
```
`err2` here?

christarazi replied:
Looks like the sole [failure](https://jenkins.cilium.io/job/Cilium-PR-K8s-1.16-kernel-4.9/1815/testReport/junit/Suite-k8s-1/16/K8sDatapathConfig_Etcd_Check_connectivity/) in CI is hidden here with `err=nil` in the sysdump because of this typo.

joestringer replied:
Ah too bad. I'll fix it up, hopefully the failure is reliable enough that it triggers next time and prints the error.

christarazi commented on pkg/k8s/watchers/endpoint.go:
```diff
@@ -156,5 +155,6 @@ func (k *K8sWatcher) addKubeAPIServerServiceEPs(ep *slim_corev1.Endpoints) {
 	}
 
-	k.handleKubeAPIServerServiceEPChanges(desiredIPs)
+	rid := ipcacheTypes.NewResourceID("ep", ep.ObjectMeta.GetNamespace(), ep.ObjectMeta.GetName())
```
I guess we'll somehow need to keep these resource types constant ("ep") so that in the future, users don't mistakenly use a different string to represent the same resource type. Defining them in the ipcache types pkg seems like the most obvious place.

joestringer replied:
The intent here is to use the canonical short name for the resource, I wonder if we already have constants for those.. Will look into it.

joestringer replied:
Ah but I take your point, there are cases that don't come from k8s. So defining these constants in ipcache package will ensure there is one canonical set of non-conflicting names. I'll put it there.

christarazi commented on pkg/k8s/watchers/endpoint.go:
```diff
@@ -105,14 +106,14 @@ func (k *K8sWatcher) deleteK8sEndpointV1(ep *slim_corev1.Endpoints, swg *lock.St
 	// by the CN handler wouldn't be overwrite-able by the entries inserted
 	// from this handler.
-	src := source.CustomResource
+	src := source.KubeAPIServer
```
Nit: comment should be updated / removed now

joestringer committed:
ipcache/md: Convert metadata key to netip.Prefix

This commit moves most of the newer ipcache logic over to "net/ipnet" so
that we can have stronger types when dealing with later additions to the
core IP structures. Previously with the string keys, the expectation was
that users of the metadata cache would insert IP addresses rather than
prefixes. However, features like CIDR policy need to insert using
prefixes. We could ambiguously insert with either IPs or CIDRs like the
main IPCache structure does, but this can easily introduce subtle bugs.
Instead, swap out the key for the ipcache.metadata structure for
netip.Prefix to force use of prefixes as the key even for existing use
cases that are based around IP addresses. As a side benefit, netip can
be used as map keys and it's supposed to be more efficient than the
prior net types.

This commit does not currently swap the main IPCache structure over to
net/ipnet, as this would force us to deal with conflicts between IP and
prefix keys in the main map. We can first introduce this in the metadata
cache, then introduce a newer IPCache metadata API that handles
conflicts (upcoming commit), migrate the existing codepaths that inject
into the ipcache to the new API, and then finally migrate the main
IPCache structure over to net/netip types.

Co-authored-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

christarazi commented on pkg/ipcache/metadata.go:
```diff
@@ -104,114 +151,178 @@ func (m *metadata) get(prefix string) labels.Labels {
+			// being used, then no need to update it.
+			if id.ID == newID.ID {
+				goto releaseIdentity
+			}
```
Isn't that what `isNew` can tell us? `isNew` comes from the second return from `injectLabels` which we are ignoring currently.

Also, is it necessary to have this logic inside the IDMD? I'm thinking why can't we just `UpsertLabels` from the kvstore side and have the IDMD calculate that it doesn't need to do anything because we have the same information and from a lower prio source?

joestringer replied:
I don't think `isNew` is quite the same. Let's take the case where we have an IP -> `remote-node` mapping. Then kube-apiserver becomes scheduled on that node. We will insert this information to the metadata map and trigger this function. When attempting to determine the new identity, it should be updated to the reserved kube-apiserver identity (7). In this case, the identity is a reserved identity so `isNew` is false. However, we do expect that the identity is different from before and the ipcache datapath should be updated. Checking `isNew` here would prevent this from updating correctly.

We'll need to be a bit careful when we extend this code to handle encryption keys better though, since we could end up having the same labels but need to update the encryption key. Maybe the generalized version of this check is "check the current BPF ipcache value and compare to the new BPF ipcache value after this change is processed, then only push the ipcache entry into the map it if it changed". The current code structure doesn't exactly facilitate this though.

I'll have to think a bit more about the latter question. One concern is that if we have multiple sources calling `UpsertLabels()`, source should *not* be considered for determining the value of the entries any more. Part of the point of the PR is to allow multiple sources of labels to be combined. There's also the question of where we write the logic to determine exactly how these updates are pushed to the ipcache. Currently we emit events for all changes into the metadata map, and all logic for processing those changes is in this function. It may make sense to push an optimization pass to avoid calls into this function via such checks at some point. Right now I'm not too sure it will make a big difference, especially given that these identities are allocated locally anyway.

christarazi replied:
> In this case, the identity is a reserved identity so isNew is false. However, we do expect that the identity is different from before and the ipcache datapath should be updated. Checking isNew here would prevent this from updating correctly.

Ah yes, I forgot about this :+1: 

> generalized version of this check is "check the current BPF ipcache value and compare to the new BPF ipcache value after this change is processed, then only push the ipcache entry into the map it if it changed"

Agree, this is likely what we need to do to associate aux information with IPs. 

---

For the latter, I was thinking a diff like below. Haven't tested it, given that Jenkins is not quite working at the moment due to missing Go 1.18 on the nodes.

```diff
diff --git a/pkg/ipcache/kvstore.go b/pkg/ipcache/kvstore.go
index e327b06d19..b3564e518a 100644
--- a/pkg/ipcache/kvstore.go
+++ b/pkg/ipcache/kvstore.go
@@ -17,7 +17,9 @@ import (
 	"github.com/sirupsen/logrus"
 
 	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache/types"
 	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/labels"
 	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/option"
@@ -221,6 +223,7 @@ func NewIPIdentityWatcher(ipc *IPCache, backend kvstore.BackendOperations) *IPId
 func (iw *IPIdentityWatcher) Watch(ctx context.Context) {
 
 	var scopedLog *logrus.Entry
+	rid := types.NewResourceID(types.ResourceKindKVStore, "", "")
 restart:
 	watcher := iw.backend.ListAndWatch(ctx, "endpointIPWatcher", IPIdentitiesPath, 512)
 
@@ -295,15 +298,15 @@ restart:
 					}
 				}
 
-				peerIdentity := ipIDPair.ID
-				if option.Config.EnableRemoteNodeIdentity && peerIdentity == identity.ReservedIdentityHost {
+				peerLabel := labels.LabelHost
+				if option.Config.EnableRemoteNodeIdentity && ipIDPair.ID == identity.ReservedIdentityHost {
 					// The only way we can discover IPs associated with the local host
 					// is directly via the NodeDiscovery package. If someone is informing
 					// this agent about IPs corresponding to the "host" via the kvstore,
 					// then they're sharing their own perspective on their own node IPs'
 					// identity. However, this node has remote-node enabled, so we should
 					// treat the peer as a "remote-node", not a "host".
-					peerIdentity = identity.ReservedIdentityRemoteNode
+					peerLabel = labels.LabelRemoteNode
 				}
 
 				// There is no need to delete the "old" IP addresses from this
@@ -312,10 +315,7 @@ restart:
 				// and the endpoint-runIPIdentitySync where it bounded to a
 				// lease and a controller which is stopped/removed when the
 				// endpoint is gone.
-				iw.ipcache.Upsert(ip, ipIDPair.HostIP, ipIDPair.Key, k8sMeta, Identity{
-					ID:     peerIdentity,
-					Source: source.KVStore,
-				})
+				iw.ipcache.UpsertLabels(ip, peerLabel, source.KVStore, rid)
 
 			case kvstore.EventTypeDelete:
 				// Value is not present in deletion event;
diff --git a/pkg/ipcache/types/types.go b/pkg/ipcache/types/types.go
index b3aeb106e9..701f0f1a95 100644
--- a/pkg/ipcache/types/types.go
+++ b/pkg/ipcache/types/types.go
@@ -41,6 +41,7 @@ var (
 	ResourceKindEndpointSlice        = ResourceKind("endpointslices")
 	ResourceKindEndpointSlicev1beta1 = ResourceKind("endpointslices_v1beta1")
 	ResourceKindNode                 = ResourceKind("node")
+	ResourceKindKVStore              = ResourceKind("kvstore")
 )
 
 // NewResourceID returns a ResourceID populated with the standard fields for
```

joestringer replied:
AFAIK ipcache/kvstore.go will receive both node and pod address/identity pairs, so we can't just assume that the identity is for a node. Specifically this diff will ignore any other identity incorrectly:

```
-				peerIdentity := ipIDPair.ID
...
+				peerLabel := labels.LabelHost
```

joestringer replied:
I think you're hoping to turn the kvstore notifications for other nodes into label metadata in the metadata cache so that it properly merges with kube-apiserver label? Maybe then we can do something special like this specifically for the host/remote-node identities.

joestringer replied:
I have also been thinking about the pod identities and the way they should be inserted into the ipcache package overall. Background assumptions are: policy either selects endpoints by label (ie select the global identity corresponding to that endpoint), or selects peers outside the cluster (typically CIDR or FQDN policy statements, where each node comes up with its own local identity for each peer/prefix).

Given these assumptions, it is possible for a conflict between the identity associated with the IP itself vs. a local identity allocated for the same IP via CIDR/FQDN policy. Given that the former policy only applies to pods and the latter is only intended to apply to connections to peers outside the cluster, we treat the presence of a global identity for the IP as a strong signal that the peer actually lives inside the cluster. If it's inside the cluster, then the CIDR policy doesn't apply. Therefore we can ignore the local identity when there is a conflict.

Today we implement this through this weird prioritization where we allow users of IPCache to insert entries into the ipcache either using an IP directly (implying it's actually a pod) or using a prefix (which could be a /32, hence directly conflicting). Then implicitly if there is a conflict like this, we choose the identity associated with the "w.x.y.z" entry rather than the "w.x.y.z/N" key in the ipcache userspace map.

We could do a similar thing instead by inserting the prefix+identity info into the metadata map directly for both cases, since the new metadata format allows us to store both types of info concurrently in the same structure. When we attempt to resolve the identity for that IP as part of the regular inner loop, first check whether any of the resources provide an identity directly, and then if so, pick that identity (prioritizing the highest source first). This could be done explicitly to make the resolution rules clear. Then only if there are no resources directly providing the identity, we fall back to the current code to merge labels from all resources and create a local identity.

I don't think we have to tackle this problem in the current PR but it should be a nice incremental step on top which could lead to a point where all ipcache interactions come in using the new interface, resolved consistently using the one clear definition on how to resolve the "correct identity" for each prefix.

christarazi replied:
Ah good call on the incorrect kvstore patch for pod identities. 

> I think you're hoping to turn the kvstore notifications for other nodes into label metadata in the metadata cache so that it properly merges with kube-apiserver label? Maybe then we can do something special like this specifically for the host/remote-node identities.

Yes this is what I was thinking.

I've updated the diff below. It still feels a bit clunky but I'm not sure there's a way around that. 

<details> 
  <summary>Diff</summary>

```diff
diff --git a/pkg/ipcache/kvstore.go b/pkg/ipcache/kvstore.go
index e327b06d19..2ac2e77b7e 100644
--- a/pkg/ipcache/kvstore.go
+++ b/pkg/ipcache/kvstore.go
@@ -17,7 +17,9 @@ import (
 	"github.com/sirupsen/logrus"
 
 	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache/types"
 	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/labels"
 	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/option"
@@ -221,6 +223,7 @@ func NewIPIdentityWatcher(ipc *IPCache, backend kvstore.BackendOperations) *IPId
 func (iw *IPIdentityWatcher) Watch(ctx context.Context) {
 
 	var scopedLog *logrus.Entry
+	rid := types.NewResourceID(types.ResourceKindKVStore, "", "")
 restart:
 	watcher := iw.backend.ListAndWatch(ctx, "endpointIPWatcher", IPIdentitiesPath, 512)
 
@@ -295,28 +298,35 @@ restart:
 					}
 				}
 
-				peerIdentity := ipIDPair.ID
-				if option.Config.EnableRemoteNodeIdentity && peerIdentity == identity.ReservedIdentityHost {
-					// The only way we can discover IPs associated with the local host
-					// is directly via the NodeDiscovery package. If someone is informing
-					// this agent about IPs corresponding to the "host" via the kvstore,
-					// then they're sharing their own perspective on their own node IPs'
-					// identity. However, this node has remote-node enabled, so we should
-					// treat the peer as a "remote-node", not a "host".
-					peerIdentity = identity.ReservedIdentityRemoteNode
+				var nodePeerLabel labels.Labels
+				if ipIDPair.ID == identity.ReservedIdentityHost || ipIDPair.ID == identity.ReservedIdentityRemoteNode {
+					nodePeerLabel = labels.LabelHost
+					if option.Config.EnableRemoteNodeIdentity && ipIDPair.ID == identity.ReservedIdentityHost {
+						// The only way we can discover IPs associated with the
+						// local host is directly via the NodeDiscovery
+						// package. If someone is informing this agent about
+						// IPs corresponding to the "host" via the kvstore,
+						// then they're sharing their own perspective on their
+						// own node IPs' identity. However, this node has
+						// remote-node enabled, so we should treat the peer as
+						// a "remote-node", not a "host".
+						nodePeerLabel = labels.LabelRemoteNode
+					}
+					// There is no need to delete the "old" IP addresses from
+					// this ip ID pair. The only places where the ip ID pair
+					// are created is the clustermesh, where it sends a delete
+					// to the KVStore, and the endpoint-runIPIdentitySync where
+					// it bounded to a lease and a controller which is
+					// stopped/removed when the endpoint is gone.
+					iw.ipcache.UpsertLabels(ip, peerLabel, source.KVStore, rid)
+				} else {
+					// Identity might be a pod identity
+					iw.ipcache.Upsert(ip, ipIDPair.HostIP, ipIDPair.Key, k8sMeta, Identity{
+						ID:     ipIDPair.ID,
+						Source: source.KVStore,
+					})
 				}
 
-				// There is no need to delete the "old" IP addresses from this
-				// ip ID pair. The only places where the ip ID pair are created
-				// is the clustermesh, where it sends a delete to the KVStore,
-				// and the endpoint-runIPIdentitySync where it bounded to a
-				// lease and a controller which is stopped/removed when the
-				// endpoint is gone.
-				iw.ipcache.Upsert(ip, ipIDPair.HostIP, ipIDPair.Key, k8sMeta, Identity{
-					ID:     peerIdentity,
-					Source: source.KVStore,
-				})
-
 			case kvstore.EventTypeDelete:
 				// Value is not present in deletion event;
 				// need to convert kvstore key to IP.
diff --git a/pkg/ipcache/types/types.go b/pkg/ipcache/types/types.go
index b3aeb106e9..701f0f1a95 100644
--- a/pkg/ipcache/types/types.go
+++ b/pkg/ipcache/types/types.go
@@ -41,6 +41,7 @@ var (
 	ResourceKindEndpointSlice        = ResourceKind("endpointslices")
 	ResourceKindEndpointSlicev1beta1 = ResourceKind("endpointslices_v1beta1")
 	ResourceKindNode                 = ResourceKind("node")
+	ResourceKindKVStore              = ResourceKind("kvstore")
 )
 
 // NewResourceID returns a ResourceID populated with the standard fields for

``` 
</details>

christarazi replied:
> We could do a similar thing instead by inserting the prefix+identity info into the metadata map directly for both cases, since the new metadata format allows us to store both types of info concurrently in the same structure. When we attempt to resolve the identity for that IP as part of the regular inner loop, first check whether any of the resources provide an identity directly, and then if so, pick that identity (prioritizing the highest source first). This could be done explicitly to make the resolution rules clear. Then only if there are no resources directly providing the identity, we fall back to the current code to merge labels from all resources and create a local identity.
>
> I don't think we have to tackle this problem in the current PR but it should be a nice incremental step on top which could lead to a point where all ipcache interactions come in using the new interface, resolved consistently using the one clear definition on how to resolve the "correct identity" for each prefix.

That sounds good to me :+1:

gandro commented on pkg/ipcache/metadata.go:
```diff
@@ -159,111 +167,165 @@ func (m *metadata) get(prefix string) prefixInfo {
+			// from different sources, for instance in etcd mode
+			// where node information is propagated both via the
+			// kvstore and via the k8s control plane. If the new
+			// security identity is the same as the noe currently
```
tiny nit:
```suggestion
			// security identity is the same as the one currently
```

gandro reviewed (approved):
I am very excited to see this land. I reviewed this mostly also to understand the new mechanism and this looks excellent to me. Thanks for the detailed comments and commit descriptions, this made reviewing a pleasure.

I'm not deeply familiar with the old mechanism, so please treat this as a medium-to-low confidence review.

christarazi commented on pkg/node/manager/manager_test.go:
```diff
@@ -73,20 +74,43 @@ func newIPcacheMock() *ipcacheMock {
 	}
 }
 
+func AddrOrPrefixToIP(ip string) (net.IP, error) {
```
Could this be useful to move to netip helpers?

joestringer replied:
I looked back over this, and this is not using the `netip` structures. Since it's based on the older type and only used in this testing package until we migrate this testing code over to `netip`, I think it is not worth migrating to a more central location.

christarazi reviewed (approved):
:tada:

tommyp1ckles commented on pkg/ipcache/types.go:
```diff
@@ -0,0 +1,81 @@
+// In an ideal world, we would use Constraints here but as of Go 1.18, these
+// cannot be used in conjunction with methods, which is how the information
+// gets injected into the IPCache.
+type IPMetadata interface{}
```
nit: this could be `type IPMetadata any`

tommyp1ckles replied:
idea: what if this was a "mergable" type interface that had behavior for mutating the resourceInfo, that way we avoid any type ambiguity.

example:
```go
type IPMetadata interface {
    Merge(*resourceInfo, source.Source)
    Unmerge(*resourceInfo, source.Source)
}

func (m *resourceInfo) merge(info IPMetadata, src source.Source) {
    info.Merge(m, src)
}

type labelMetadataImpl labels.Labels

func (l labelMetadataImpl) Merge(m *resourceInfo, src source.Source) {
	nl := labels.NewLabelsFromModel(nil)
	nl.MergeLabels(l)
	m.labels = nl
}
```

(not an actual review point, just brainstorming)

joestringer replied:
One of my goals here is to move all of the "merging" code into this one file, right next to each other, so it can be easily compared side-by-side. If we start adding constraints onto the `IPMetadata` like requiring methods to merge/unmerge, then we will go back to a point where the code for merging / handling conflicts / resolving priority is spread across all of the callers elsewhere in the codebase. That makes it much easier to miss some newly introduced combination.

joestringer replied:
I picked up the `interface{}` -> `any` since it looks a bit tidier. Looks like it compiles and should be functioanlly identical :+1:

tommyp1ckles replied:
Yeah that's true, having the callers implement their own methods to satisfy the interface wouldn't be great. I was thinking you could just warp/alias all the necessary types here, kind of like with labelMetadataImpl in my example, that way all your merge logic is in one place.

I only bring this up because I usually try to avoid type switching/asserting where possible (unless it's a container type). It kind of goes against using a static typing system imo.

joestringer replied:
I see yeah, we could do that. But then if we end up warp/aliasing then it seems like we will need to end up doing the type switch/assert anyway?

I do think Go generics could eventually solve this for us by declaring it like this:

```
type IPMetadata interface{
    labels.Labels | foo | bar
}
```

However, I couldn't get this to work with the current implementation of go generics. Maybe at some point, but this is also not critical to solve this way.

tommyp1ckles commented on pkg/ipcache/types.go:
```diff
@@ -0,0 +1,81 @@
+		l.MergeLabels(info)
+		m.labels = l
+	default:
+		log.Errorf("BUG: Invalid IPMetadata passed to ipinfo.merge(): %+v", info)
```
Should this return in this case to avoid setting `m.source = src`?

joestringer replied:
What I would _like_ here is actually a compile-time failure. I'm not sure the best way to do this though.. I tried playing around with generics but that wouldn't work because these are methods defined against some receiver, not functions. Go generics only work for functions AFAIK.

I think you're probably right that it's better to avoid merging any data here in case it impacts something else without actually implementing this merge. :+1: I'll add a return statement here.

tommyp1ckles commented on pkg/ipcache/metadata.go:
```diff
@@ -56,37 +75,43 @@ type metadata struct {
+		}
+	}
 
-	m.Lock()
```
We don't need this lock anymore?

joestringer replied:
Good catch, I'll take a closer look. I'm sure my thinking here was that this is now an internal-only function that is only called when it's locked. But at the same time, the common style for writing such a function in Cilium would be to call this `upsertLocked()` if the callers should have already grabbed the lock. That's not the case, so I probably missed something during the refactor.

christarazi replied:
It looks like we might need the locking inside `UpsertMetadata` and rename `upsert` to `upsertLocked`.

joestringer replied:
It looks like there was just one caller here and it was handling the locking correctly. But I've renamed the function to make the locking pattern more clear. :+1:

joestringer committed:
ipcache: Introduce new asynchronous API

Introduce a new generic API for associating information with the
IPs in the IPCache, which accounts for multiple sources of information
such as labels coming from different resources (eg Services ->
reserved:kube-apiserver, NetPol -> CIDR labels).

The primary core of this API is the UpsertMetadata(...) function, which
takes the following parameters:
- prefix: IP (range) that this info applies to;
- src: info source of the information;
- resource: specific resource name in the information source,
- aux: variable length list of information to associate with the prefix.

'aux' is typed as IPMetadata, which is effectively just an interface{}
to allow any information to be associated with the IPCache. Developers
should read the comments around IPMetadata and expand the IPCache
package in the relevant areas (particularly pkg/ipcache/types.go and the
InjectLabels() call) to ensure that the IPCache package correctly
handles the new information and effectively merges the different sources
of info correctly.

After info is upserted into the IPCache via this new API, it will
automatically trigger an out-of-band resolution of what the new IPCache
entry for the prefix should look like, taking into account each piece of
source information from various resources.

In this patch, we switch the current kube-apiserver logic over to the
new API as an initial example, removing the need for the caller to
trigger label injection since the new API will automatically schedule
this. Future work will expand this to switch other subsystems over to
the new APIs, introducing new resourceInfo fields and merging logic in
the ipcache package to decide how complementary (or even conflicting)
information should be combined in order to generate IPCache entries.

Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>
